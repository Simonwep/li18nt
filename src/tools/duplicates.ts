import {JSONArray, JSONObject, PropertyPath} from '@types';
import {typeOfJsonValue} from '@utils/typeOfJsonValue';

/**
 * Finds duplicate keys in the given object.
 * @param object
 */
export const duplicates = (object: JSONObject): Map<string, PropertyPath[]> => {
    const duplicates: Map<string, PropertyPath[]> = new Map<string, PropertyPath[]>();
    const keys: string[] = [];

    const walk = (entry: JSONObject | JSONArray, parentPath: PropertyPath): void => {
        if (Array.isArray(entry)) {
            for (let i = 0; i < entry.length; i++) {
                const value = entry[i];

                if (typeOfJsonValue(value) === 'object') {
                    walk(value as JSONObject, [...parentPath, i]);
                }
            }
        } else {
            for (const key in entry) {
                if (Object.prototype.hasOwnProperty.call(entry, key)) {
                    const value = entry[key];
                    const type = typeOfJsonValue(value);

                    if (type === 'object' || type === 'array') {
                        walk(value as JSONObject, [...parentPath, key]);
                    } else if (keys.includes(key)) {
                        const list = duplicates.get(key) || [];
                        duplicates.set(key, [...list, [...parentPath, key]]);
                    } else {
                        keys.push(key);
                    }
                }
            }
        }
    };

    walk(object, []);
    return duplicates;
};
