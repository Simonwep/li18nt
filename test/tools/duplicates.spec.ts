import {duplicates} from '@tools/duplicates';

describe('[tool] duplicates', () => {

    it('Should find a duplicate property and its paths', () => {
        expect(duplicates({
            'welcome': 'Willkommen',
            'home': {
                'welcome': 'Willkommen'
            }
        })).toMatchSnapshot();
    });

    it('Should find multiple, nested duplicate', () => {
        expect(duplicates({
            'welcome': 'Willkommen',
            'back': 'Zurück',
            'home': {
                'welcome': 'Willkommen',
                'sideBar': {
                    'welcome': 'Willkommen'
                }
            }
        })).toMatchSnapshot();
    });

    it('Should work with arrays too', () => {
        expect(duplicates({
            'welcome': 'Willkommen',
            'copyright': 'Urheberrecht',
            'back': 'Zurück',
            'home': {
                'welcome': 'Willkommen',
                'sideBar': {
                    'welcome': 'Willkommen'
                }
            },
            'about': {
                'footer': [
                    {'welcome': 'Willkommen'},
                    {'welcome': 'Willkommen'},
                    {'welcome': 'Willkommen', 'copyright': 'Urheberrecht'}
                ]
            }
        })).toMatchSnapshot();
    });
});
