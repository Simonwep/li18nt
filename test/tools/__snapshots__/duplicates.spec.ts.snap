// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[tool] duplicates Should find a duplicate property and its paths 1`] = `
Map {
  "welcome" => Array [
    Array [
      "home",
      "welcome",
    ],
  ],
}
`;

exports[`[tool] duplicates Should find multiple, nested duplicate 1`] = `
Map {
  "welcome" => Array [
    Array [
      "home",
      "welcome",
    ],
    Array [
      "home",
      "sideBar",
      "welcome",
    ],
  ],
}
`;

exports[`[tool] duplicates Should work with arrays too 1`] = `
Map {
  "welcome" => Array [
    Array [
      "home",
      "welcome",
    ],
    Array [
      "home",
      "sideBar",
      "welcome",
    ],
    Array [
      "about",
      "footer",
      0,
      "welcome",
    ],
    Array [
      "about",
      "footer",
      1,
      "welcome",
    ],
    Array [
      "about",
      "footer",
      2,
      "welcome",
    ],
  ],
  "copyright" => Array [
    Array [
      "about",
      "footer",
      2,
      "copyright",
    ],
  ],
}
`;
